### PR TITLE
Store fee policy

### PR DESCRIPTION
**Pull request by sunce86**
# Description
Fixes https://github.com/cowprotocol/services/issues/2111

Defines the table for storing fee policies per `<auction_id, order_uid>`. 

Currently supports taking fees as:
1. cut from a surplus - the price difference between the executed price and limit_price. Optionally, for this type, we also have a CAP on max volume (expressed in the same way as (2)
2. percent of volume (or more precise, percent of the `executed_amount`, so refers to sell token for sell orders and buy token for buy orders).

Will be merged at the very end of feature implementation, since db migrations are irreversible.

# Changes

- [x] Created table for fee policies
- [x] Implemented basic database function to read/write

## How to test
Roundtrip UT

### CODE REVIEW

File: crates/database/src/fee_policies.rs
Comment (fleupold) [inline crates/database/src/fee_policies.rs:28]: What's the advantage of properly structuring the kind into columns? I think this will just make it very challenging to migrate. Would it make sense to instead store json-ified versions of the fee policy so we can add new policies without a db migration?

File: crates/database/src/fee_policies.rs
Comment (sunce86) [reply crates/database/src/fee_policies.rs:36]: My reasoning to define it here is so that it could be reusable (potentially) by both orderbook and autopilot. Autopilot and ordebook should probably have their own dto's for this structure.

File: database/sql/V057__add_fee_policy.sql
Comment (MartinquaXD) [inline database/sql/V057__add_fee_policy.sql:10]: The name `insertion_order` seems off. Couldn't the items be inserted in any order? The only thing that matters is the order in which the policies get applied. So maybe `application_order` or just `index`? 🤷‍♂️

File: crates/autopilot/src/arguments.rs
Comment (fleupold) [inline crates/autopilot/src/arguments.rs:306]: Could you rebase the PR to only show the things that are new to this PR?

File: crates/database/src/fee_policies.rs
Comment (fleupold) [reply crates/database/src/fee_policies.rs:28]: @MartinquaXD what do you think here? I don't fully get the argument about db space (I'm not sure storing 3 values is more/less space efficient than a potentially compressed json blob, but in any case the difference is likely not meaningful for us).

I think we should think about maintainability (ie as we define more fee types is it better to have them as potentially formatted jsons or do we want to migrate the DB on each change (I don't really foresee advantages of the querying capacities, ie. I don't think we need to e.g. query all orders with absolute fee > some value)

**Review by fleupold — COMMENTED** [review]
Hard to review currently, since it's mixing 2 big PRs

File: database/sql/V057__add_fee_policy.sql
Comment (MartinquaXD) [reply database/sql/V057__add_fee_policy.sql:3]: > I wouldn't document the details of the different fee policies inside the database

We have a DB [readme](https://github.com/cowprotocol/services/blob/main/database/README.md) for some time now which should be up-to-date. Let's keep it that way.

File: crates/database/src/fee_policies.rs
Comment (MartinquaXD) [reply crates/database/src/fee_policies.rs:28]: > So the advantage would be that we use less db space, since we have an entry for each auction for each order.

We only really need to store the data for orders that were settled in an auction, right?
In that case we should rather the format that fits our needs best and add a clean up task that deletes data that is no longer needed.

> I think we should think about maintainability (ie as we define more fee types is it better to have them as potentially formatted jsons or do we want to migrate the DB on each change

This would be the reason for me why preferring regular columns over JSON blobs make sense. One can easily imagine a scenario where you would want to look up some historic data which no longer works because the JSON format changed in an incompatible way in the mean time. When we have real columns we are forced to more conscious about keeping the system coherent. That way we can still chose to phase out old incompatible formats without being surprised all of a sudden.

Comment (fleupold) [reply crates/database/src/fee_policies.rs:28]: Right, but FeePolicy to me is something very "atomic" on the order which we are not likely to sub-query for specific ranges (buy I might by wrong).
In this case, I'd strongly argue to not overload the semantics of each column though (if volume fees have a volume % and quote deviation orders have a volume fee cap in %, let's define separate columns for both). No developer in their right mind is going to understand in 6 months time under what circumstances a value in one column affects the semantics of a value in another column.

Comment (sunce86) [reply crates/database/src/fee_policies.rs:28]: Switched to storing all columns and not overloading a single column for multiple things.

File: crates/database/src/fee_policies.rs
Comment (sunce86) [reply crates/database/src/fee_policies.rs:91]: dropped code

File: database/sql/V057__add_fee_policy.sql
Comment (sunce86) [reply database/sql/V057__add_fee_policy.sql:3]: removed comments

File: database/sql/V057__add_fee_policy.sql
Comment (sunce86) [reply database/sql/V057__add_fee_policy.sql:17]: removed.

File: database/sql/V057__add_fee_policy.sql
Comment (sunce86) [reply database/sql/V057__add_fee_policy.sql:10]: index is reserved term. switched to `application_order`

File: database/sql/V057__add_fee_policy.sql
Comment (sunce86) [reply database/sql/V057__add_fee_policy.sql:13]: The column can store numbers with up to 3 digits in total, 2 of which can be after the decimal point

File: crates/database/src/fee_policies.rs
Comment (sunce86) [reply crates/database/src/fee_policies.rs:41]: Kept the "user friendly" common type because it encapsulates how the fields are populated for different policy types (IMO autopilot and orderbook should not worry about it).

File: crates/database/src/fee_policies.rs
Comment (fleupold) [inline crates/database/src/fee_policies.rs:35]: why lowercae? Didn't this make composite words hard to read? What's our standard naming policy here?

File: crates/database/src/fee_policies.rs
Comment (fleupold) [reply crates/database/src/fee_policies.rs:26]: Isn't "factor" for the volume variant still missing (or implicitly overloading "volume factor")?

Just as a thought experiment, how would we go about adding a new FeePolicyKind variant later on?

File: database/sql/V057__add_fee_policy.sql
Comment (fleupold) [inline database/sql/V057__add_fee_policy.sql:14]: What's the purpose of these checks? I think it we want to enforce invariants, we should do so in the rust code and not rely on database specific syntax (to me the fee policy database could simply be a key value store that allows looking up the fee policy for a specific order).

**Review by fleupold — COMMENTED** [review]
cc @alfetopito as this is lying the foundation of how fee policies for orders can later be returned to the frontend.

@sunce86 have we mapped out how this process will work?

File: crates/database/src/fee_policies.rs
Comment (sunce86) [reply crates/database/src/fee_policies.rs:35]: We use lowercase for database. Used the same here for consistency.

File: database/sql/V057__add_fee_policy.sql
Comment (sunce86) [reply database/sql/V057__add_fee_policy.sql]: Thought it was nice to have it since already available out of the box. Removed.

File: crates/database/src/fee_policies.rs
Comment (sunce86) [reply crates/database/src/fee_policies.rs:26]: Updated.

> Just as a thought experiment, how would we go about adding a new FeePolicyKind variant later on?

We would add a new enum variant to the kind and added new columns specific to this kind.

**Comment by sunce86** [comment]
> @sunce86 have we mapped out how this process will work?

Currently we don't have a requirement to allow frontend to fetch policies right now. 

Fetching fee policies together with order can be done by joining on the `fee_policies` table.

File: database/sql/V057__add_fee_policy.sql
Comment (fleupold) [inline database/sql/V057__add_fee_policy.sql:7]: Is SERIAL correct here? Shouldn't this just be the index of the fee policy for the given (auction, order) pair?

File: crates/autopilot/src/database/competition.rs
Comment (fleupold) [inline crates/autopilot/src/database/competition.rs:71]: Given the two things are orthogonal, can we make this a separate method, `save_fee_policies` (and maybe start adding the fee policy persistence into the `infra` module)

Comment (squadgazzz) [reply crates/autopilot/src/database/competition.rs:71]: Ok, will do. But this still should be a part of the same DB transaction, right? I was worried about the inconsistency.

Comment (fleupold) [reply crates/autopilot/src/database/competition.rs:71]: The data that is being written is completely different though, isn't it? Other than performance, is there any advantage of sharing the transaction?

File: database/sql/V057__add_fee_policy.sql
Comment (sunce86) [reply database/sql/V057__add_fee_policy.sql:7]: I've put SERIAL so that it would autoincrement every time we insert the same `<auction_id, order_uid>`. How would you version with a simple index look like?

Comment (squadgazzz) [reply database/sql/V057__add_fee_policy.sql:7]: auction_id | order_uid | application_order
-- | -- | --
1 | 1 | 1
1 | 1 | 2
2 | 2 | 3

should probably become

auction_id | order_uid | application_order
-- | -- | --
1 | 1 | 1
1 | 1 | 2
2 | 2 | 1

That would require manual indexing within the (auction, order). Correct me if I am wrong.

File: crates/autopilot/src/database/competition.rs
Comment (squadgazzz) [reply crates/autopilot/src/database/competition.rs]: Migrated to infra

Comment (sunce86) [reply crates/autopilot/src/database/competition.rs]: We should probably first merge https://github.com/cowprotocol/services/pull/2188 since a lot of refactoring is done there so no point moving any of the functionality from this PR to `infra` without first looking into ☝️ PR.

**Comment by harisang** [comment]
Had a discussion with Dusan today and, although you have discussed this and Dusan explained you have good reasons for going this way, having an order whose fee policy can change depending on which auction it got settled in is strange. As a user, when I sign an order, I would like to know what kind of protocol fee i will pay. Implying that depending on what the core team or the DAO will choose, I might pay some fee that I have no idea how it will look like, doesn't sound reasonable to me.

Anyway, not a deal breaker, but it is something that, I, as a user, would definitely complain about (especially since users have expressed that they want even longer expiration times, for example, for limit orders; in such a case, suggesting that now the fee is 20% of surplus capped, but maybe in 6 months it will be 100% of surplus uncapped is problematic)

**Comment by fleupold** [comment]
> As a user, when I sign an order, I would like to know what kind of protocol fee i will pay. Implying that depending on what the core team or the DAO will choose, I might pay some fee that I have no idea how it will look like, doesn't sound reasonable to me.

cc @olgafetisova. It doesn't really relate to this PR specifically, but is more a general thought on how the fee is designed (and the question whether an order's fee policy is allowed to change over time).

File: database/sql/V057__add_fee_policy.sql
Comment (squadgazzz) [reply database/sql/V057__add_fee_policy.sql:7]: Manual indexing adds more complexity to the code. @fleupold, let me know if I got the intention correctly, so I'll be able to continue with that or just skip it.

Comment (sunce86) [reply database/sql/V057__add_fee_policy.sql:7]: I'm fine with both (although the second one is prettier). What is important is that the ordering is right. Meaning, given a pair <auction_id, order_id> give me all the policies ordered the same way as they were applied (application_order)

Comment (sunce86) [reply database/sql/V057__add_fee_policy.sql:7]: I think the first option also gives us an opportunity to get ordered policies just by a given order (not even using auction_ids). Not sure if we going to need this option though.

Comment (fleupold) [reply database/sql/V057__add_fee_policy.sql:7]: I see (I thought option 2 was the intended one, but it's true that 1 works as well). No strong feelings, whatever is simpler to implement.

File: crates/autopilot/src/run_loop.rs
Comment (fleupold) [inline crates/autopilot/src/run_loop.rs:129]: Not needed after rebase since policies are now part of the auction

File: crates/autopilot/src/infra/persistence/fee_policy/mod.rs
Comment (fleupold) [inline crates/autopilot/src/infra/persistence/fee_policy/mod.rs:8]: Combine with existing impl

File: crates/autopilot/src/infra/persistence/fee_policy/mod.rs
Comment (fleupold) [inline crates/autopilot/src/infra/persistence/fee_policy/mod.rs:18]: If this is the dto for storing in the db, do we then need an extra `FeePolicyRow` struct inside the postgres impl? Or should the dto be what we can right straight into the db?

File: crates/autopilot/src/infra/persistence/fee_policy/mod.rs
Comment (fleupold) [inline crates/autopilot/src/infra/persistence/fee_policy/mod.rs:9]: Interface should probably take the domain::auction and the order_uid iterator for matched orders only and then fetch the auction ID and fee policies from that.

File: crates/autopilot/src/infra/persistence/fee_policy/mod.rs
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/fee_policy/mod.rs:18]: The `database` crate should not have any dependencies, so the `FeePolicyRow` should remain. Do I miss anything?

**Comment by sunce86** [comment]
Explicitly copying this comment here to take into consideration when defining which structs to use for fee policy and how it should be shared: https://github.com/cowprotocol/services/pull/2250#discussion_r1444628774

File: crates/autopilot/src/infra/persistence/fee_policy/mod.rs
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/fee_policy/mod.rs]: The FeePolicy -> FeePolicyRow conversion contains some internal logic that should not be visible through the DTO. But we can remove the additional layer for sure.

File: crates/autopilot/src/run_loop.rs
Comment (squadgazzz) [inline crates/autopilot/src/run_loop.rs:286]: Decided to store them separately from the `Competition` in order to properly use metrics since this is not a critical part of the flow.

File: crates/autopilot/src/infra/persistence/mod.rs
Comment (sunce86) [inline crates/autopilot/src/infra/persistence/mod.rs:104]: `auction_id: domain::AuctionId`

File: crates/autopilot/src/infra/persistence/mod.rs
Comment (sunce86) [inline crates/autopilot/src/infra/persistence/mod.rs:126]: can we move dto conversion code to `crates/autopilot/src/infra/persistence/dto/fee_policies.rs`

File: crates/database/src/fee_policies.rs
Comment (sunce86) [inline crates/database/src/fee_policies.rs:8]: can we kill this struct and just implement conversion from `autopilot::domain::fee::Policy` to `FeePolicyRow` directly and put that into `crates/autopilot/src/infra/persistence/dto/fee_policy.rs`

File: crates/database/src/fee_policies.rs
Comment (sunce86) [inline crates/database/src/fee_policies.rs:81]: similar to other comment, adjust this to convert to domain fee policy object and move it to autopilot infra.

File: crates/database/src/fee_policies.rs
Comment (squadgazzz) [reply crates/database/src/fee_policies.rs:8]: Sure, we can, but then we will have only `domain::fee::Policy` without `auction_id` and `order_uid` fields.

File: crates/autopilot/src/run_loop.rs
Comment (sunce86) [inline crates/autopilot/src/run_loop.rs:180]: I would push the fee policy only if the solution is a winner solution. There is a code below that similarly checks for winner: 

`let is_winner = solutions.len() - index == 1;`

Otherwise, if two or more solvers return solutions containing the same order, we will end up with multiple same fee policies for the same <auction_id, order_uid> and based on the current implementation those would be different entries in the table with different `application_order` column AFAIS.

Comment (squadgazzz) [reply crates/autopilot/src/run_loop.rs:180]: But currently, we are working with the latest solution.

Comment (sunce86) [reply crates/autopilot/src/run_loop.rs:180]: Oh, my bad. All good.

File: crates/autopilot/src/infra/persistence/fee_policy/mod.rs
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/fee_policy/mod.rs]: Everything was fixed after the merge

File: crates/database/src/fee_policies.rs
Comment (MartinquaXD) [inline crates/database/src/fee_policies.rs:25]: You could take a `PgConnection` here since `PgTransaction` automatically dereferences into that type. Unless you want to enforce that you can only use this function in a transaction.

File: database/sql/V057__add_fee_policy.sql
Comment (MartinquaXD) [inline database/sql/V057__add_fee_policy.sql:18]: This automatically creates an index, right?
If we select by `auction_id` and `order_id` but **NOT** by `application_order` would we still hit this index or would we have to create an index with `(auction_id, order_id)`?

File: crates/autopilot/src/run_loop.rs
Comment (MartinquaXD) [inline crates/autopilot/src/run_loop.rs:281]: This is on the hot path (i.e. can limit our throughput by delaying the start of the next auction).
Since this information is pretty important we should probably not offload that to a background task (in case that fails) but we should optimize performance of that call by parallelizing it or preferably having a single batch insert.

File: database/sql/V057__add_fee_policy.sql
Comment (MartinquaXD) [inline database/sql/V057__add_fee_policy.sql:1]: We have a [README](https://github.com/cowprotocol/services/blob/main/database/README.md) for the database.
Would be good to make a habit out of updating that whenever we add a new DB migration.

File: database/sql/V057__add_fee_policy.sql
Comment (squadgazzz) [reply database/sql/V057__add_fee_policy.sql:18]: The PK index is used while querying by `(auction_id, order_id)`. Adding a separate index here would probably be sufficient if we expect a lot of queries. Otherwise, it makes sense to keep it as it is to avoid additional disk usage.

File: crates/autopilot/src/run_loop.rs
Comment (squadgazzz) [reply crates/autopilot/src/run_loop.rs:180]: True, thanks, since solution contains unique `order_uids`

File: database/sql/V057__add_fee_policy.sql
Comment (sunce86) [reply database/sql/V057__add_fee_policy.sql:18]: We have a test `postgres_roundtrip` that shows that by specifying the `(auction_id, order_id)` all rows are returned which is a desired behavior.

File: crates/database/src/fee_policies.rs
Comment (squadgazzz) [reply crates/database/src/fee_policies.rs]: Fixed

File: crates/autopilot/src/run_loop.rs
Comment (squadgazzz) [reply crates/autopilot/src/run_loop.rs:276]: Done. I genually tried to avoid [this](https://github.com/cowprotocol/services/blob/3aaa3d91232ce7999204fd5c40c2e844c9ac45d5/crates/autopilot/src/infra/persistence/mod.rs#L117) and [this](https://github.com/cowprotocol/services/blob/3aaa3d91232ce7999204fd5c40c2e844c9ac45d5/crates/autopilot/src/infra/persistence/mod.rs#L120) by using something like `Iter<Item = IntoIter<Item = FeePolicyRow>>`, but couldn't find a proper way to make it work with async function.

File: database/sql/V057__add_fee_policy.sql
Comment (squadgazzz) [reply database/sql/V057__add_fee_policy.sql:1]: https://github.com/cowprotocol/services/pull/2118/commits/22f279fc34c7ffbbd7421ca393427e81b5a892a3

File: crates/autopilot/src/arguments.rs
Comment (sunce86) [inline crates/autopilot/src/arguments.rs:61]: Can we move this down to be part of the `FeePolicy` group of parameters.

`pub fee_policy: FeePolicy,`

File: crates/autopilot/src/run_loop.rs
Comment (MartinquaXD) [reply crates/autopilot/src/run_loop.rs:276]: Yeah, iterators and async functions can get pretty hairy due to the references. Optimizing this to the fullest is probably not worth the headache.
One small thing that could be done is to omit the allocation of another vector for each batch with:
```rust
            database::fee_policies::insert_batch(&mut ex, chunk.iter().cloned())
```

**Review by MartinquaXD — APPROVED** [review]

File: crates/autopilot/src/arguments.rs
Comment (squadgazzz) [reply crates/autopilot/src/arguments.rs]: Done

File: crates/autopilot/src/arguments.rs
Comment (fleupold) [inline crates/autopilot/src/arguments.rs:369]: Do we foresee any reason for why this value should be different from `order_events_insert_batch_size`? If so, can we combine them in a database related struct (which can be flattened on the top args, but forms a place to but db related configs, maybe also with other configs) or rename `order_events_insert_batch_size` and reuse it?

File: crates/autopilot/src/run_loop.rs
Comment (fleupold) [inline crates/autopilot/src/run_loop.rs:280]: If we cannot recover the fee policies later aren't we running into trouble from an accounting perspective? I think we should potentially not continue the run-loop in this case. Wdyt?

File: crates/database/src/fee_policies.rs
Comment (fleupold) [inline crates/database/src/fee_policies.rs:1]: Why should this live in the shared database crate vs. autopilot::database?

File: crates/autopilot/src/infra/persistence/dto/fee_policy.rs
Comment (fleupold) [inline crates/autopilot/src/infra/persistence/dto/fee_policy.rs:9]: Can the `FeePolicyRow` struct be lifted up here or would this cause other weird dependencies? I find it weird to have a `dto::fee_policy` model that is not really defining any dto.

Also, I'd prefer this to be available via `database::FeePolicy` or `dto::FeePolicy` rather than `FeePolicyRow` (while this naming is consistent with the database crate, it doesn't really fit into the new DDD types, maybe we can re-export it somehow?)

**Review by fleupold — APPROVED** [review]
Looks good overall. If we are going to rename the fee policy type, should we do this before we seed any initial types into the database (otherwise we may run into unnecessary migration issues)

File: crates/autopilot/src/arguments.rs
Comment (squadgazzz) [reply crates/autopilot/src/arguments.rs:369]: Different values could be theoretically required since data size differs for a single row in different tables. For now,  we can probably combine them into a single config param to simplify things.

File: crates/autopilot/src/run_loop.rs
Comment (squadgazzz) [reply crates/autopilot/src/run_loop.rs:280]: I think @sunce86 mentioned that we should not stop the loop just because of the fee policies storing error since the information is not so important. Is this information outdated already?

File: crates/database/src/fee_policies.rs
Comment (squadgazzz) [reply crates/database/src/fee_policies.rs:1]: Is it expected that the data will be fetched from the orderbook later?

File: crates/autopilot/src/run_loop.rs
Comment (sunce86) [reply crates/autopilot/src/run_loop.rs:280]: Autopilot/protocol is responsible for determining and storing/retrieving fee policies for orders. In that context, if it failed for some orders/auctions I believe it is ok to consider (in accounting) that we took 0 protocol fees for those orders/auctions.

I am in favor of keeping it as is (for now) since this is a new feature and there might be bugs. I would hate to have a prod downtime caused by protocol fee bug. Once stable, I would probably switch to more strict version (skip runloop).

File: crates/autopilot/src/infra/persistence/dto/fee_policy.rs
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/dto/fee_policy.rs]: Moved together with the `database::fee_policies`.

**Pull request merged**
